### PR TITLE
[Java][Go] Unalias schema in getTypeDeclaration when not generating model for alias

### LINKED ISSUE #4362: [BUG][Java][Go-experimental] Referenced schema for `array` type badly generated
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

Defining a schema for an array type, and referencing it as the `items` type in another array doesn't generate valid code. It happened for me in Java and go-experimental that I'm using, but might be the case for other languages as well.

##### openapi-generator version

4.1.3

##### OpenAPI declaration file content or url

The following schemas definition exhibit the behavior 
```yaml
NestedArray:
  type: array
  items:
    type: string
MainSchema:
  type: object
  properties:
    nested_array:
      type: array
      items:
        $ref: '#/components/schemas/NestedArray'
``` 

In go, the generated output looks like this, which doesn't even compile since `Array` is undefined:
```go
type MainSchema struct {
	NestedArray *[]Array `json:"nested_array,omitempty"`
}
```

instead of the expected output of either
```go
type MainSchema struct {
	NestedArray *[]NestedArray `json:"nested_array,omitempty"`
}
```
if we generated a model for the `NestedArray` schema, or simply
```go
type MainSchema struct {
	NestedArray *[][]string `json:"nested_array,omitempty"`
}
```
if we don't generate a model (which it seems is what is supposed to happen when we create a schema for a primitive type).

In java, the output is
```java
public class MainSchema {
  public static final String JSON_PROPERTY_NESTED_ARRAY = "nested_array";
  private List<List> nestedArray = null;
}
```
instead of 
```java
public class MainSchema {
  public static final String JSON_PROPERTY_NESTED_ARRAY = "nested_array";
  private List<List<String>> nestedArray = null;
}
```

Please note that the correct output is generated when inlining the `NestedArray` definition instead of referencing it in the `MainSchema`, like so:
```yaml
MainSchema:
  type: object
  properties:
    nested_array:
      type: array
      items:
        type: array
        items:
          type: string
``` 
##### Command line used for generation

openapi-generator generate --http-user-agent DataDog/1.0.0/java -g java -c config/languages/java_v1.json -i spec/v1/full_spec.yaml -o generated/datadog-api-client-java/v1 -t templates/java'

##### Steps to reproduce

Make a spec with the above schema definitions, and inspect the generated model for `MainSchema`

### PR DESCRIPTION
**Pull request by zippolyte**
This is my attempt to fix #4362 for the two languages that I'm having this issue with.
What do you think ?

@bbdouglas @sreeshas @jfiala @lukoyanov @cbornet @jeff9finger @karismann @Zomzog @lwlee2608

@antihax @bvwells @grokify @kemokemo @bkabrda 

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

**Comment by wing328** [comment]
@zippolyte thanks for the PR. I think it's better to fix it in the DefaultCodegen. I'll try to fix it later this week.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractGoCodegen.java [hunk 1]
```diff
@@ -278,10 +278,16 @@ public String getTypeDeclaration(Schema p) {
         if (ModelUtils.isArraySchema(p)) {
             ArraySchema ap = (ArraySchema) p;
             Schema inner = ap.getItems();
-            return "[]" + getTypeDeclaration(inner);
+            if (ModelUtils.isGenerateAliasAsModel())
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractGoCodegen.java]: @zippolyte I don't think you need `ModelUtils.isGenerateAliasAsModel()` as `unaliasSchema` has already taken that into account:  https://github.com/OpenAPITools/openapi-generator/blob/10627ed5cc3f4703b0f02662ea568353e8d26e00/modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java#L812

**Comment by wing328** [comment]
UPDATE: I've revised your fix and I think it's the right approach. I've left a comment suggesting a minor change. Please review when you've time. Thanks again for the PR.

**Comment by zippolyte** [comment]
Indeed it seems I don't need that after all. Thanks for your review, updating.

**Comment by wing328** [comment]
@zippolyte thanks for the PR, which has been merged into master. I've filed https://github.com/OpenAPITools/openapi-generator/issues/4563 to track the fixes for other generators.

**Comment by wing328** [comment]
@zippolyte thanks for the PR, which has been included in the v4.2.2 release: https://twitter.com/oas_generator/status/1201432648544972800

**Pull request merged**